### PR TITLE
Hw1 cleancode

### PR DESCRIPTION
**Pull request by Miha-s**

### CODE REVIEW

File: internal/emails_storage/emails_storage_impl.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+type EmailsStorageImpl struct {
+	emails            map[string]struct{}
+	storage_file_path string
+	storage_name      string
+}
+
+func fileExists(filepath string) bool {
+	info, err := os.Stat(filepath)
+	if os.IsNotExist(err) {
```
Comment (CalculusEnjoyer) [inline internal/emails_storage/emails_storage_impl.go:19]: Would be nice to add some common errors in a separate file with friendly error messages

**Review by CalculusEnjoyer — COMMENTED** [review]
Also, you may add email some simple email validator to filter inaccurate emails

File: internal/lifecycle/lifecycle.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	return nil
+}
+
+func (lifecycle *Lifecycle) Run() error {
+	done := make(chan error, 1)
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt, syscall.SIGTERM)
+
+	defer lifecycle.services.EmailStorage.Close()
```
[the remark this answers is not in the document]
Comment (Miha-s) [reply internal/lifecycle/lifecycle.go:73]: Solid comment, will try fix

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+version: '3.4'
+services:
+  btc-test-task:
+    build: .
+    image: btc-test-task
+    ports:
+      - '2777:2777'
+    expose:
```
[the remark this answers is not in the document]
Comment (Miha-s) [reply docker-compose.yml:8]: I suppose you are right

File: internal/config/config.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	CurrencyTo              string
+}
+
+func (conf *Config) LoadFromENV() error {
+	err := godotenv.Load(".env")
+	if err != nil {
+		return err
+	}
+
+	port, _ := strconv.Atoi(os.Getenv("PORT"))
```
[the remark this answers is not in the document]
Comment (Miha-s) [reply internal/config/config.go:30]: Agree

File: internal/rate_accessors/coinapi.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	api.api_key = conf.CoinAPIKey
+	return nil
+}
+
+func extractRate(json_value []byte) (float64, error) {
+	var dat map[string]interface{}
+	if err := json.Unmarshal(json_value, &dat); err != nil {
+		return 0, err
+	}
+	return dat["rate"].(float64), nil
```
[the remark this answers is not in the document]
Comment (Miha-s) [reply internal/rate_accessors/coinapi.go:28]: Okay... will add

File: internal/rate_accessors/coinapi.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	api_key  string
+}
+
+func (api *CoinApI) Init(conf *config.Config) error {
+	api.endpoint = conf.CoinAPIUrl + conf.CurrencyFrom + "/" + conf.CurrencyTo
+	api.api_key = conf.CoinAPIKey
+	return nil
+}
+
+func extractRate(json_value []byte) (float64, error) {
```
[the remark this answers is not in the document]
Comment (Miha-s) [reply internal/rate_accessors/coinapi.go:23]: I will think about it

File: internal/emails_storage/emails_storage_impl.go (hunk 1 above)
Comment (Miha-s) [reply internal/emails_storage/emails_storage_impl.go:19]: Good observation, I were going to refactor error handling

File: internal/email_sender/email_sender_impl.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	gomail "gopkg.in/gomail.v2"
+)
+
+type EmailSenderImpl struct {
+	email   string
+	subject string
+	dialer  gomail.Dialer
+}
+
+func (sender *EmailSenderImpl) Init(conf *config.Config) error {
```
Comment (andrey-filenko-quarks-tech) [inline internal/email_sender/email_sender_impl.go:18]: I guess it should be changed to constructor method, so the user of this struct would not have to manually use .Init function, cause it creates some kind of invariant, where you can have unitialized EmailSender

File: internal/emails_storage/emails_storage_impl.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+	for key := range *set {
+		result = append(result, key)
+	}
+
+	return result
+}
+
+func (storage *EmailsStorageImpl) Close() {
+	logger.LogInfo("Closing file storage")
+	storage_file, err := os.Create(storage.storage_file_path)
```
Comment (andrey-filenko-quarks-tech) [inline internal/emails_storage/emails_storage_impl.go:47]: variable naming should in pascalCase, without undescore

File: internal/emails_storage/emails_storage_impl.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+	storage.storage_name = "emails_storage.json"
+	storage.emails = make(map[string]struct{})
+	storage.storage_file_path = conf.EmailStoragePath + "/" + storage.storage_name
+	if !fileExists(storage.storage_file_path) {
+		return nil
+	}
+	return storage.openExistingStorage()
+}
+
+func getArrayFromSet(set *map[string]struct{}) []string {
```
Comment (andrey-filenko-quarks-tech) [inline internal/emails_storage/emails_storage_impl.go:35]: why pointer?

File: internal/lifecycle/lifecycle.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Miha-s) [reply internal/lifecycle/lifecycle.go:71]: fixed

File: docker-compose.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Miha-s) [reply docker-compose.yml]: fixed

File: internal/config/config.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Miha-s) [reply internal/config/config.go]: fixed

File: internal/rate_accessors/coinapi.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Miha-s) [reply internal/rate_accessors/coinapi.go]: fixed

File: internal/rate_accessors/coinapi.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Miha-s) [reply internal/rate_accessors/coinapi.go]: fixed

File: internal/emails_storage/emails_storage_impl.go (hunk 1 above)
Comment (Miha-s) [reply internal/emails_storage/emails_storage_impl.go]: fixed

File: internal/email_sender/email_sender_impl.go (hunk 7 above)
Comment (Miha-s) [reply internal/email_sender/email_sender_impl.go]: You are right, done

File: internal/email_sender/email_sender_impl.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+import (
+	"btc-test-task/internal/config"
+	"btc-test-task/internal/logger"
+
+	"crypto/tls"
+
+	gomail "gopkg.in/gomail.v2"
+)
+
+type EmailSenderImpl struct {
```
[the remark this answers is not in the document]
Comment (Miha-s) [reply internal/email_sender/email_sender_impl.go]: Yes, sounds much better, done

File: internal/emails_storage/emails_storage_impl.go (hunk 8 above)
Comment (Miha-s) [reply internal/emails_storage/emails_storage_impl.go]: Understood, done

File: internal/emails_storage/emails_storage_impl.go (hunk 9 above)
Comment (Miha-s) [reply internal/emails_storage/emails_storage_impl.go]: Fixed

**Review by CalculusEnjoyer — APPROVED** [review]
Good job, bro 😎🤗

**Pull request merged**